File: BEmu/ReferenceDataRequest/ReferenceRequest.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="ReferenceDataRequest/ReferenceRequest.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ReferenceRequest : Request
    {
        private readonly ReferenceRequestElementStringArray _securities, _fields;
        private readonly ReferenceRequestElementOverrideArray _overrides;

        public ReferenceRequest()
        {
            this._securities = new ReferenceRequestElementStringArray("securities");
            this._fields = new ReferenceRequestElementStringArray("fields");
            this._overrides = new ReferenceRequestElementOverrideArray();
        }

        internal List<string> Securities { get { return this._securities.Values; } }
        internal List<string> Fields { get { return this._fields.Values; } }

        public override IEnumerable<Element> Elements
        {
            get
            {
                yield return this._securities;
                yield return this._fields;

                if (this._overrides.NumValues > 0)
                    yield return this._overrides;
            }
        }

        public override bool HasElement(string name)
        {
            if (this._securities.Name.ToString() == name)
                return this._securities.Values.Any();
            else if (this._fields.Name.ToString() == name)
                return this._fields.Values.Any();
            else if (this._overrides.Name.ToString() == name)
                return this._overrides.NumValues > 0;
            else
                return false;
        }

        public override Element this[string elementName]
        {
            get
            {
                if (this._securities.Name.ToString() == elementName)
                    return this._securities;
                else if (this._fields.Name.ToString() == elementName)
                    return this._fields;
                else if (this._overrides.Name.ToString() == elementName)
                    return this._overrides;
                else
                    throw new NotImplementedException("BEmu.ReferenceDataRequest.RequestReference: public Element this[string elementName] not supported.  ElementName is case-sensitive");
            }
        }

        public override void Append(string name, string elementValue)
        {
            switch (name)
            {
                case "securities":
                    this._securities.AppendValue(elementValue);
                    break;
                case "fields":
                    this._fields.AppendValue(elementValue);
                    break;
                default:
                    throw new ArgumentException(string.Format("BEmu.RequestReference.Append: Element name {0} not supported. Names are case-sensitive", name));
            }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.AppendFormat("ReferenceDataRequest = {{{0}", Environment.NewLine);

            if (this._securities.NumValues > 0)
                result.Append(this._securities.PrettyPrint(1));

            if (this._fields.NumValues > 0)
                result.Append(this._fields.PrettyPrint(1));

            if (this._overrides.NumValues > 0)
                result.Append(this._overrides.PrettyPrint(1));

            result.AppendLine("}");
            return result.ToString();
        }

    }
}

File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceRequest.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceRequest.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Request;
import com.bloomberglp.blpapi.Element;
import java.util.List;

public class ReferenceRequest extends Request
{
    private final ReferenceRequestElementStringArray _securities, _fields;
    private final ReferenceRequestElementOverrideArray _overrides;

    public ReferenceRequest()
    {
        this._securities = new ReferenceRequestElementStringArray("securities");
        this._fields = new ReferenceRequestElementStringArray("fields");
        this._overrides = new ReferenceRequestElementOverrideArray();
    }
    
    List<String> securities()
    {
    	return this._securities.values();
    }
    
    List<String> fields()
    {
    	return this._fields.values();
    }
    
    public Element getElement(String name) throws Exception
    {
        if (this._securities.name().toString().equals(name))
            return this._securities;
        
        else if (this._fields.name().toString().equals(name))
            return this._fields;
        
        else if (this._overrides.name().toString().equals(name))
            return this._overrides;
            
        else
            return super.getElement(name);
    }
    
    public boolean hasElement(String name) throws Exception
    {
        if (this._securities.name().toString().equals(name))
            return this._securities.values().size() > 0;
        
        else if (this._fields.name().toString().equals(name))
            return this._fields.values().size() > 0;
        
        else if (this._overrides.name().toString().equals(name))
            return this._overrides.numValues() > 0;
            
        else
            return false;
    }
    
    public void append(String name, String elementValue) throws Exception
    {
    	if(name.equals("securities"))
    		this._securities.appendValue(elementValue);
    	
    	else if(name.equals("fields"))
    		this._fields.appendValue(elementValue);
    	
    	else
    		throw new Exception(String.format("BEmu.RequestReference.Append: Element name %s not supported. names are case-sensitive.", name));
    }
    
    public String toString()
    {
        StringBuilder result = new StringBuilder();
        result.append("ReferenceDataRequest = {" + System.getProperty("line.separator"));

        if (this._securities.numValues() > 0)
            result.append(this._securities.prettyPrint(1));

        if (this._fields.numValues() > 0)
            result.append(this._fields.prettyPrint(1));

        try
        {
			if (this._overrides.numValues() > 0)
			    result.append(this._overrides.prettyPrint(1));
		}
        catch (Exception e) { }

        result.append("}" + System.getProperty("line.separator"));

        return result.toString();
    }
    
}
